src/pgpdisk_fmt_plug.c

```diff
@@ -0,0 +1,369 @@
+		out[i] = (atoi16[ARCH_INDEX(*p)] << 4) | atoi16[ARCH_INDEX(p[1])];
+		p += 2;
+	}
+
+	return out;
+}
+
+#undef SHA1_DIGEST_LENGTH
+#define SHA1_DIGEST_LENGTH 20
+
+// HashSaltSchedulePassphrase in original source code
+static void pgpdisk_kdf(char *password, unsigned char *salt, unsigned char *key, int key_length)
+{
+	uint32_t bytesNeeded = key_length;
+	uint32_t offset = 0;
+	unsigned char hash[SHA1_DIGEST_LENGTH];
+	int plen;
+	int iterations = cur_salt->iterations;
+	SHA_CTX ctx; // SHA1 usage is hardcoded
+
+	plen = strlen(password);
+	while (bytesNeeded > 0) {
+		uint32_t bytesThisTime = SHA1_DIGEST_LENGTH < bytesNeeded ? SHA1_DIGEST_LENGTH: bytesNeeded;
+		uint8_t j;
+		uint16_t i;
+
+		SHA1_Init(&ctx);
+		if (offset > 0) {
+			SHA1_Update(&ctx, key, SHA1_DIGEST_LENGTH);
+		}
+		SHA1_Update(&ctx, password, plen);
+		SHA1_Final(hash, &ctx);
+
+		SHA1_Init(&ctx);
+		if (cur_salt->algorithm == 3)
+			SHA1_Update(&ctx, salt, 8); // kNumSaltBytes = 8, for CAST5
+		else
+			SHA1_Update(&ctx, salt, 16); // kNumSaltBytes = 16, for AES-256, Twofish
+
+		for (i = 0, j = 0; i < iterations; i++, j++) {
+			SHA1_Update(&ctx, hash, bytesThisTime);
+			SHA1_Update(&ctx, &j, 1);
+		}
+		SHA1_Final(key + offset, &ctx);
+
+		bytesNeeded -= bytesThisTime;
+		offset += bytesThisTime;
+	}
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+
+#ifdef _OPENMP
+#pragma omp parallel for
+	for (index = 0; index < count; index += MAX_KEYS_PER_CRYPT)
+#endif
+	{
+		int i;
+
+		for (i = 0; i < MAX_KEYS_PER_CRYPT; i++) {
+			unsigned char key[40];
+
+			if (cur_salt->algorithm == 5) {
+				AES_KEY aes_key;
+
+				pgpdisk_kdf(saved_key[i+index], cur_salt->salt, key, 32);
+				// DecryptPassphraseKey in original source code, compute CheckBytes
+				AES_set_encrypt_key(key, 256, &aes_key);
+				AES_ecb_encrypt(key, (unsigned char*)crypt_out[index+i], &aes_key, AES_ENCRYPT);
+			} else if (cur_salt->algorithm == 4) {
+				Twofish_key tkey;
+
+				pgpdisk_kdf(saved_key[i+index], cur_salt->salt, key, 32);
+				Twofish_prepare_key(key, 32, &tkey);
+				Twofish_encrypt(&tkey, key, (unsigned char*)crypt_out[index+i]);
+			} else if (cur_salt->algorithm == 3) {
+				CAST_KEY ck;
+
+				pgpdisk_kdf(saved_key[i+index], cur_salt->salt, key, 16);
+				CAST_set_key(&ck, 16, key);
+				memset((unsigned char*)crypt_out[index+i], 0, BINARY_SIZE);
+				CAST_ecb_encrypt(key, (unsigned char*)crypt_out[index+i], &ck, CAST_ENCRYPT);
+			}
+		}
+	}
+
+	return count;
+}
+
+static int cmp_all(void *binary, int count)
+{
+	int index = 0;
+	for (; index < count; index++)
+		if (((uint32_t*)binary)[0] == crypt_out[index][0])
+			return 1;
+	return 0;
+}
+
+static int cmp_one(void *binary, int index)
+{
+	return !memcmp(binary, crypt_out[index], BINARY_SIZE);
+}
+
+static int cmp_exact(char *source, int index)
+{
+	return 1;
+}
+
+static void set_key(char *key, int index)
+{
+	strnzcpyn(saved_key[index], key, sizeof(saved_key[index]));
```

Comment:
Nit-pick: Use `strnzcpy()` (with no trailing "n") when you don't need the length in return. When you do, something like `saved_len[index] = strnzcpyn(saved_key[index], key, sizeof(saved_key[index]));` can be relevant. This latter version is jumbo-specific IIRC.

Not that it matters much. I guess you lose a quarter of a billionth of a second or so 😆